canonical/tempo-coordinator-k8s-operator
# load-balance between scaled-up coordinator units

## Issue
When the coordinator is scaled up, this doesn't reflect in the ingress configuration of `traefik_route`. So, requests to traefik will always be proxied to one unit of the coordinator although we might have multiple units in place.

## Solution
Listen on `peer_relation_created` hook to populate peer data with hostnames of each coordinator unit and when constructing the ingress configuration, we retrieve the servers urls from the peer data.

 ## Drive-bys
 - Remove any DNS reference in the worker config to the coordinator or traefik DNS names and depend solely on the workers' DNS names. The rationale behind that is that the worker config sent to a worker should indicate that this config is sufficient and we can run a worker process with no need for sth outside the worker to be ready. i.e: If we pass traefik's hostname as `tls_server_name` in the worker config, at any point in time, traefik can now be behind TLS and our worker will suddenly fail as it needs Traefik's CA..
 - Update `CertHandler` lib
 
 ## Testing Considerations
 1. deploy this bundle
 ```
bundle: kubernetes
applications: 
  ssc:
    charm: self-signed-certificates
    channel: latest/edge
    trust: true
  minio:
    charm: minio
    channel: latest/edge
    options:
      access-key: accesskey
      secret-key: mysoverysecretkey
    trust: true
  traefik:
    charm: traefik-k8s
    channel: latest/edge
    scale: 1
    constraints: arch=amd64
    trust: true
  tempo:
    charm: /home/michael/Work/tempo-coordinator-k8s-operator/tempo-coordinator-k8s_ubuntu-22.04-amd64.charm
    scale: 3
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo,anti-pod.topology-key=kubernetes.io/hostname
    trust: true
    resources:
      nginx-image: ubuntu/nginx:1.24-24.04_beta
      nginx-prometheus-exporter-image: nginx/nginx-prometheus-exporter:1.1.0
  s3-integrator:
    charm: s3-integrator
    channel: latest/edge
    scale: 1
    constraints: arch=amd64
  tempo-compactor:
    charm: tempo-worker-k8s
    channel: latest/edge
    resources:
      tempo-image: 3
    scale: 3
    options:
      role-all: false
      role-compactor: true
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo-compactor,anti-pod.topology-key=kubernetes.io/hostname
    storage:
      data: kubernetes,1,1024M
    trust: true
  tempo-distributor:
    charm: tempo-worker-k8s
    channel: latest/edge
    resources:
      tempo-image: 3
    scale: 3
    options:
      role-all: false
      role-distributor: true
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo-distributor,anti-pod.topology-key=kubernetes.io/hostname
    storage:
      data: kubernetes,1,1024M
    trust: true
  tempo-ingester:
    charm: tempo-worker-k8s
    channel: latest/edge
    resources:
      tempo-image: 3
    scale: 3
    options:
      role-all: false
      role-ingester: true
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo-ingester,anti-pod.topology-key=kubernetes.io/hostname
    storage:
      data: kubernetes,1,1024M
    trust: true
  tempo-metrics-generator:
    charm: tempo-worker-k8s
    channel: latest/edge
    resources:
      tempo-image: 3
    scale: 3
    options:
      role-all: false
      role-metrics-generator: true
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo-metrics-generator,anti-pod.topology-key=kubernetes.io/hostname
    storage:
      data: kubernetes,1,1024M
    trust: true
  tempo-querier:
    charm: tempo-worker-k8s
    channel: latest/edge
    resources:
      tempo-image: 3
    scale: 3
    options:
      role-all: false
      role-querier: true
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo-querier,anti-pod.topology-key=kubernetes.io/hostname
    storage:
      data: kubernetes,1,1024M
    trust: true
  tempo-query-frontend:
    charm: tempo-worker-k8s
    channel: latest/edge
    resources:
      tempo-image: 3
    scale: 3
    options:
      role-all: false
      role-query-frontend: true
    # constraints: arch=amd64 tags=anti-pod.app.kubernetes.io/name=tempo-query-frontend,anti-pod.topology-key=kubernetes.io/hostname
    storage:
      data: kubernetes,1,1024M
    trust: true
```
 2. run `juju run s3-integrator/leader sync-s3-credentials access-key=accesskey secret-key=mysoverysecretkey`
 3. create a `tempo` bucket
 4. run `juju config s3-integrator endpoint=minio-0.minio-endpoints.test.svc.cluster.local:9000 bucket=tempo`
 5. run `jhack imatrix fill`
 6. wait units settle down (except for metrics-generator). With TLS, it will take around 30 mins to settle
 7. run `juju ssh --container nginx tempo/leader 'nginx -s stop'`. You might want to run that at least 7 times to have some backoff before nginx auto restarts.
 8. run `curl <INGRESS_URL>:3200/api/search` 3 times
 9. You should be able to see the traces in 2/3 curls. 1/3 curls would return a `Bad Gateway` as its routed to the down coordinator instance.
